service/matching/taskListManager.go

```diff
@@ -474,22 +578,55 @@ func (c *taskListManagerImpl) trySyncMatch(task *persistence.TaskInfo) (*persist
 	}
 	// Request from the point of view of Add(Activity|Decision)Task operation.
 	// But it is getTask result from the point of view of a poll operation.
-	request := &getTaskResult{task: task, C: make(chan *syncMatchResponse, 1)}
+	request := &getTaskResult{task: task, C: make(chan *syncMatchResponse, 1), syncMatch: true}
+
+	rsv := c.rateLimiter.Reserve()
+	if !rsv.OK() {
+		scope := metrics.MatchingTaskListMgrScope
+		c.metricsClient.IncCounter(scope, metrics.AddThrottleCounter)
+		return nil, fmt.Errorf("cannot add to tasklist, limit exceeded")
```

Comment:
Are you suggesting returning nil here and not an error at all? Or returning a specific error type  that is handled in the handler code. For the second option, we would add a new error type in thrift, and Max was a bit determined that this should not be considered an error which is why I also named the metric a counter instead of an error. Some history from previous thread with Max when I had used a ServiceBusyError classification for this:

mfateev on Nov 28, 2017  Member
I don't think we want to return service busy error. Also it is going to be returned even if there is no backlog as we throttle before trying to match. From the poller point of view it is ErrNoTasks.

      @madhuravi
madhuravi on Nov 28, 2017  Member
Same as above, want to differentiate between truly no tasks vs tasks being throttled. Either way I need a new thrift type so user can react to it. ErrNewTasks is not a type, it just has a specific string message.

Reply:
Pushed a change for this. Let me know what you think.